tests/test_discriminated_union.py

```diff
@@ -1232,3 +1232,28 @@ class TestModel(BaseModel):
         'title': 'TestModel',
         'type': 'object',
     }
+
+
+def test_function_after_discriminator():
+    class CatModel(BaseModel):
+        name: Literal['kitty', 'cat']
+
+        @field_validator('name', mode='after')
+        def replace_name(cls, v):
+            return 'cat'
+
+    class DogModel(BaseModel):
+        name: Literal['puppy', 'dog']
+
+        # comment out the 2 field validators and model will work!
+        @field_validator('name', mode='after')
+        def replace_name(cls, v):
+            return 'dog'
+
+    AllowedAnimal = Annotated[Union[CatModel, DogModel], Field(discriminator='name')]
+
+    class Model(BaseModel):
+        x: AllowedAnimal
+
+    m = Model(x={'name': 'kitty'})
+    assert m.x.name == 'cat'
```

Comment:
Might be worth confirming a discriminated union was used, either via an error case or a json schema (or both)

Reply:
I was relying on the docs tests for that but I'll add it here too